pkg/networkservice/common/timeout/server.go

```diff
@@ -21,80 +21,145 @@ import (
+	"github.com/networkservicemesh/sdk/pkg/tools/log"
 )
 
 type timeoutServer struct {
-	onTimeout   *networkservice.NetworkServiceServer
-	connections map[string]*time.Timer
-	executor    serialize.Executor
+	ctx         context.Context
+	connections timerMap
+	executors   executorMap
 }
 
-// NewServer - creates a new NetworkServiceServer chain element that implements timeout of expired connections
-//             - onTimeout - *networkservice.NetworkServiceServer.  Since networkservice.NetworkServiceServer is an interface
-//                        (and thus a pointer) *networkservice.NetworkServiceServer is a double pointer.  Meaning it
-//                        points to a place that points to a place that implements networkservice.NetworkServiceServer
-//                        This is done because when we use timeout.NewServer as part of a chain, we may not *have*
-//                        a pointer to this server used 'onTimeout'.  If we detect we need to heal, onHeal.Request is used to heal.
-//                        If onTimeout is nil, then we simply set onTimeout to this server chain element
-//                        If we are part of a larger chain, we should pass the resulting chain into
-//                        this constructor before we actually have a pointer to it.
-func NewServer(onTimout *networkservice.NetworkServiceServer) networkservice.NetworkServiceServer {
-	rv := &timeoutServer{
-		connections: make(map[string]*time.Timer),
-		onTimeout:   onTimout,
-	}
-	if rv.onTimeout == nil {
-		var actualOnTimeout networkservice.NetworkServiceServer = rv
-		rv.onTimeout = &actualOnTimeout
-	}
-	return rv
+type timer struct {
+	timer  *time.Timer
+	stopCh chan struct{}
 }
 
-func (t *timeoutServer) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (*networkservice.Connection, error) {
-	ct, err := t.createTimer(ctx, request)
-	if err != nil {
-		return nil, err
+// NewServer - creates a new NetworkServiceServer chain element that implements timeout of expired connections.
+func NewServer(ctx context.Context) networkservice.NetworkServiceServer {
+	return &timeoutServer{
+		ctx: ctx,
 	}
+}
+
+func (t *timeoutServer) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (conn *networkservice.Connection, err error) {
+	logEntry := log.Entry(ctx).WithField("timeoutServer", "Request")
+
 	connID := request.GetConnection().GetId()
-	t.executor.AsyncExec(func() {
-		if timer, ok := t.connections[connID]; !ok || timer.Stop() {
-			t.connections[connID] = ct
+
+	executor, _ := t.executors.LoadOrStore(connID, &serialize.Executor{})
+	<-executor.AsyncExec(func() {
+		// executor was possibly removed by `t.close()` at this moment, we need to store it back
+		exec, _ := t.executors.LoadOrStore(connID, executor)
+		if exec != executor {
+			err = errors.Errorf("race condition, parallel request execution: %v", connID)
+			return
+		}
+
+		if timer, ok := t.connections.Load(connID); ok {
```

Comment:
Please use LoadAndDelete

Reply:
`go-syncmap` doesn't support such method generation. Added it manually.